fix(helm): Delete hooks should wait for resource to be removed from etcd before continuing

mortent opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/master/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
There is a race condition when using the `before-hook-creation` deletion policy. The code in Tiller makes the delete call, but this only sets the deletion timestamp on the resource and doesn't actually remove the object from etcd. With finalizers, there might be some time before the object is actually removed. If the object isn't removed by the time the object is re-created, the chart will fail.
This PR adds code fixes the delete call used in the hook to wait for the object to be deleted from etcd before considering the delete operation completed.

Fixes #5375

**Special notes for your reviewer**:

**If applicable**:
- [x] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

mortent committed:
fix(helm): Delete hooks should wait for resource to be removed from etcd before continuing

Signed-off-by: Morten Torkildsen <mortent@google.com>

mortent commented:
@micw I think this should fix the issue you are describing in #5375. I'm having trouble reproducing the issue, so could you try to build Helm with this change and see if it works?

thomastaylor312 commented:
@mortent Please ping me once you are able to reproduce the issue with @micw or any other person experiencing the issue and I'll review for you

thomastaylor312 commented on pkg/kube/client.go:
```diff
@@ -391,16 +392,25 @@ func (c *Client) Update(namespace string, originalReader, targetReader io.Reader
 //
 // Namespace will set the namespace.
-func (c *Client) Delete(namespace string, reader io.Reader) error {
+func (c *Client) Delete(namespace string, reader io.Reader, timeout int64, shouldWait bool) error {
```
Just an FYI that this is a breaking API change. Not many people should be using `pkg/kube` directly that much, but if we need to add the timeout, we'll probably need to add another method so as to not break backwards compatibility

mortent commented:
I have updated the PR to add a new delete method in addition to the existing one to avoid breaking the API. I've also added unit tests.

I'm not able to reliably reproduce the reported problem on GKE, but the way the delete hook works today does have a weakness that should be addressed by this change. 

@thomastaylor312 This should be ready for review. If you have a chance, #5112 has some similarities to this one and needs review from a second maintainer.

micw commented:
@mortent @thomastaylor312 Sorry, I was a bit busy those days. I try to verify this PR in my test setup.

micw commented:
tl;dr: Works!

My procedure to reproduce it:

- clone the helm repo
- fetch the PR as new branch: git fetch origin pull/5532/head:pr-5532
- checkout the branch: git  checkout pr-5532
- build+push helm and the docker image: https://gist.github.com/micw/30190083b3e5d7df8dc30bc1526bdb57
- ran my test script from https://github.com/micw/helm-hook-test with official helm release
--> failed: Error: object is being deleted: customresourcedefinitions.apiextensions.k8s.io "tests.my.test.com" already exists

- switched tiller to micwy/helm-test:pr-5532

```
$ helm version
Client: &version.Version{SemVer:"v2.12.3", GitCommit:"eecf22f77df5f65c823aacd2dbd30ae6c65f186e", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.12+unreleased", GitCommit:"7196ba364d46f93685feb486c255b21609703196", GitTreeState:"clean"}
```

- ran my script again

... running since 10 minutes without error ;-)

I let it run this night and send an update tomorrow.

thomastaylor312 commented:
Thanks for verifying @micw!

thomastaylor312 commented on pkg/tiller/environment/environment.go:
```diff
@@ -119,6 +119,16 @@ type KubeClient interface {
 	// by "\n---\n").
 	Delete(namespace string, reader io.Reader) error
 
+	// Delete destroys one or more resources. If shouldWait is true, the function
```
Comment should start with `DeleteWithTimeout`

thomastaylor312 commented on pkg/tiller/release_server.go:
```diff
@@ -448,7 +448,7 @@ func (s *ReleaseServer) deleteHookByPolicy(h *release.Hook, policy string, name,
 	if hookHasDeletePolicy(h, policy) {
 		s.Log("deleting %s hook %s for release %s due to %q policy", hook, h.Name, name, policy)
-		if errHookDelete := kubeCli.Delete(namespace, b); errHookDelete != nil {
+		if errHookDelete := kubeCli.DeleteWithTimeout(namespace, b, 60, true); errHookDelete != nil {
```
Should this timeout value be configurable by annotation on the hook? If we don't want that, then this definitely should be a constant so we can give it a meaningful name

mortent replied:
I've moved the timeout value into a constant. Making it configurable makes sense, although I think the default should be a timeout larger than 0. Waiting for the resource to be fully deleted is a reasonable default behavior.

So would it make sense to create an additional annotation, something like `helm.sh/hook-delete-timeout` that would override the default value? Or allow appending the timeout value at the end of the current hooks, something like `before-hook-creation:60`? I think I prefer the first alternative.

thomastaylor312 replied:
I think I agree with you with what the name should be. Let me check with some of the other maintainers as well.

thomastaylor312 replied:
Yeah, let's go with `helm.sh/hook-delete-timeout` and add that as an override for the default. Thanks for the work on this @mortent!

thomastaylor312 reviewed (commented):
Just a few points of discussion

thomastaylor312 commented:
@mortent I am working on testing this now. Once you add hook-delete-timeout, let me know and I'll give it a final review!

thomastaylor312 commented:
Ok, this is working great and I confirmed that it works with the test chart @micw used to replicate the issue. As soon as we have that timeout configuration in place, this should be good to go

mortent commented:
@thomastaylor312 I've updated the PR. If it looks good, I can squash the commits before we merge.

thomastaylor312 commented:
Sorry! Somehow this totally slipped through my notifications. I'll look at it first thing tomorrow.

thomastaylor312 reviewed (approved):
This is looking great! I'll ping one of the other maintainers for a second review

hickeyma commented:
Review in progress. Testing looks good. Just need to eye ball the code next.

hickeyma reviewed (commented):
@mortent Do you mind squashing the commits?

hickeyma reviewed (changes requested):
@mortent It's nearly there. Can you update the (docs)[https://github.com/helm/helm/blob/master/docs/charts_hooks.md] with the new hook?

mortent commented:
I'll squash the commits and update the docs today.

mortent commented:
@hickeyma @thomastaylor312 I have squashed the commits and updated the docs.

hickeyma commented on pkg/kube/client.go:
```diff
@@ -449,15 +449,34 @@ func (c *Client) cleanup(newlyCreatedResources []*resource.Info) (cleanupErrors
 //
 // Namespace will set the namespace.
 func (c *Client) Delete(namespace string, reader io.Reader) error {
```
Is this function now obsolete?

thomastaylor312 replied:
I don't think it is, but even if it is obsolete, we cannot delete it because it would break backwards compatibility (since it is an exported function)

hickeyma replied:
Ok, probably best to leave it then.

hickeyma reviewed (approved):
LGTM @mortent. Thanks for all the work and changes.
